### PR TITLE
Fix `top_k` for `multiclass-f1score`

### LINKED ISSUE #1653: `top_k` for `multiclassf1score` is not working correctly
## 🐛 Bug

`top_k` argument of `MulticlassF1Score` is not working as expected. It suppose to give higher results as `top_k `increases, but that is not happening sometimes. 

According to docs: 
> top_k (int) – Number of highest probability or logit score predictions considered to find the correct label. 

So, it must increase strictly always.

Also normally, when `top_k=num_classes`, it is expected to give 1 (100%), but that's not happening either.

### To Reproduce

Steps to reproduce the behavior...

<details>
  <summary>Code sample</summary>

```python
import torch
from torchmetrics.classification import MulticlassF1Score

preds = torch.randn(200, 5).softmax(dim=-1)
target = torch.randint(5, (200,))

f1_val_top1=MulticlassF1Score(num_classes=5, top_k=1, average="macro")
f1_val_top3=MulticlassF1Score(num_classes=5, top_k=3, average="macro")
f1_val_top5=MulticlassF1Score(num_classes=5, top_k=5, average="macro")

print(f1_val_top1(preds, target), f1_val_top3(preds, target), f1_val_top5(preds, target))  
```
It returns `(tensor(0.1774), tensor(0.2740), tensor(0.3318))`

AFAI understood from documentation, when I set `top_k=5`, it must give 1 because there only 5 classes anyway.

More explicitly, I was expected the following two to have the same output:

```python
import torch, functorch
from torchmetrics.classification import MulticlassF1Score
preds = torch.randn(10, 5).softmax(dim=-1)
target = torch.randint(5, (10,))

f1_val_top3=MulticlassF1Score(num_classes=5, top_k=3, average="macro")
f1_val_top1=MulticlassF1Score(num_classes=5, top_k=1, average="macro")

pred_top_3 = torch.argsort(preds, dim=1, descending=True)[:, :3]
pred_top_1 = pred_top_3[:, 0]

# This simply changes the incorrect labels with the correct ones, only if correct guesses is in top 3 predictions 
pred_corrected_top3 = torch.where(functorch.vmap(lambda t1, t2: torch.isin(t1, t2))(target, pred_top_3), target, pred_top_1)

print(f1_val_top3(preds, target), f1_val_top1(pred_corrected_top3, target))
```
But result is different

</details>

### Environment

- TorchMetrics 0.11.3 (installed via `pip`):
- Python 3.8.16
- PyTorch 1.12.0

### PR DESCRIPTION
**Pull request by rittik9**
## What does this PR do?

Fixes #1653

<details>
  <summary>Before submitting</summary>

- [x] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
📚 Documentation preview 📚: https://torchmetrics--2839.org.readthedocs.build/en/2839/

### CODE REVIEW

File: tests/unittests/classification/test_precision_recall.py [hunk 1]
```diff
@@ -395,7 +395,7 @@ def test_multiclass_precision_recall_half_gpu(self, inputs, module, functional,
     ("k", "preds", "target", "average", "expected_prec", "expected_recall"),
     [
         (1, _mc_k_preds, _mc_k_target, "micro", tensor(2 / 3), tensor(2 / 3)),
-        (2, _mc_k_preds, _mc_k_target, "micro", tensor(1 / 2), tensor(1.0)),
+        (2, _mc_k_preds, _mc_k_target, "micro", tensor(1.0), tensor(1.0)),
```
Comment (Borda) [inline tests/unittests/classification/test_precision_recall.py:398]: let's keep the original case too?

Comment (rittik9) [reply tests/unittests/classification/test_precision_recall.py:398]: Sorry, I didn’t understand. Could you please clarify?

**Comment by rittik9** [comment]
Sure

File: tests/unittests/classification/test_precision_recall.py (hunk 1 above)
Comment (Borda) [reply tests/unittests/classification/test_precision_recall.py:398]:
```suggestion
        (2, _mc_k_preds, _mc_k_target, "micro", tensor(1 / 2), tensor(1.0)),
        (2, _mc_k_preds, _mc_k_target, "micro", tensor(1.0), tensor(1.0)),
```

File: tests/unittests/classification/test_specificity.py [hunk 2]
```diff
@@ -362,7 +362,7 @@ def test_multiclass_specificity_dtype_gpu(self, inputs, dtype):
     ("k", "preds", "target", "average", "expected_spec"),
     [
         (1, _mc_k_preds, _mc_k_target, "micro", tensor(5 / 6)),
-        (2, _mc_k_preds, _mc_k_target, "micro", tensor(1 / 2)),
+        (2, _mc_k_preds, _mc_k_target, "micro", tensor(1.0)),
```
Comment (Borda) [inline tests/unittests/classification/test_specificity.py:365]:
```suggestion
        (2, _mc_k_preds, _mc_k_target, "micro", tensor(1 / 2)),
        (2, _mc_k_preds, _mc_k_target, "micro", tensor(1.0)),
```

File: tests/unittests/classification/test_stat_scores.py [hunk 3]
```diff
@@ -369,9 +369,9 @@ def test_raises_error_on_too_many_classes(preds, target, ignore_index, error_mes
     ("k", "preds", "target", "average", "expected"),
     [
         (1, _mc_k_preds, _mc_k_target, "micro", torch.tensor([2, 1, 5, 1, 3])),
-        (2, _mc_k_preds, _mc_k_target, "micro", torch.tensor([3, 3, 3, 0, 3])),
+        (2, _mc_k_preds, _mc_k_target, "micro", torch.tensor([3, 0, 6, 0, 3])),
```
Comment (Borda) [inline tests/unittests/classification/test_stat_scores.py:372]:
```suggestion
        (2, _mc_k_preds, _mc_k_target, "micro", torch.tensor([3, 3, 3, 0, 3])),
        (2, _mc_k_preds, _mc_k_target, "micro", torch.tensor([3, 0, 6, 0, 3])),
```

File: tests/unittests/classification/test_stat_scores.py [hunk 4]
```diff
@@ -369,9 +369,9 @@ def test_raises_error_on_too_many_classes(preds, target, ignore_index, error_mes
...
         (1, _mc_k_preds, _mc_k_target, None, torch.tensor([[0, 1, 1], [0, 1, 0], [2, 1, 2], [1, 0, 0], [1, 1, 1]])),
-        (2, _mc_k_preds, _mc_k_target, None, torch.tensor([[1, 1, 1], [1, 1, 1], [1, 1, 1], [0, 0, 0], [1, 1, 1]])),
+        (2, _mc_k_preds, _mc_k_target, None, torch.tensor([[1, 1, 1], [0, 0, 0], [2, 2, 2], [0, 0, 0], [1, 1, 1]])),
```
Comment (Borda) [inline tests/unittests/classification/test_stat_scores.py:374]:
```suggestion
        (2, _mc_k_preds, _mc_k_target, None, torch.tensor([[1, 1, 1], [1, 1, 1], [1, 1, 1], [0, 0, 0], [1, 1, 1]])),
        (2, _mc_k_preds, _mc_k_target, None, torch.tensor([[1, 1, 1], [0, 0, 0], [2, 2, 2], [0, 0, 0], [1, 1, 1]])),
```

**Review by Borda — APPROVED** [review]

File: src/torchmetrics/functional/classification/stat_scores.py [hunk 5]
```diff
@@ -341,6 +341,30 @@ def _multiclass_stat_scores_format(
     return preds, target
 
 
+def refine_preds_oh(preds: Tensor, preds_oh: Tensor, target: Tensor, top_k: int) -> Tensor:
```
Comment (Borda) [inline src/torchmetrics/functional/classification/stat_scores.py:344]:
```suggestion
def _refine_preds_oh(preds: Tensor, preds_oh: Tensor, target: Tensor, top_k: int) -> Tensor:
```
let's rather make it a private function?

File: CHANGELOG.md [hunk 6]
```diff
@@ -24,10 +24,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 -
 
-
 ### Fixed
 
-- Fixed issue with shared state in metric collection when using dice score ([#2848](https://github.com/PyTorchLightning/metrics/pull/2848))
+
```
Comment (Borda) [inline CHANGELOG.md:29]:
```suggestion
- Fixed issue with shared state in metric collection when using dice score ([#2848](https://github.com/PyTorchLightning/metrics/pull/2848))


```

File: CHANGELOG.md [hunk 7]
```diff
@@ -24,12 +24,14 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 -
```
Comment (Borda) [inline CHANGELOG.md:26]:
```suggestion


```

File: tests/unittests/classification/test_stat_scores.py [hunk 8]
```diff
@@ -361,27 +362,85 @@ def test_raises_error_on_too_many_classes(preds, target, ignore_index, error_mes
...
+    [0.3, 0.3, 0.4],
+])
+
 
 @pytest.mark.parametrize(
     ("k", "preds", "target", "average", "expected"),
     [
         (1, _mc_k_preds, _mc_k_target, "micro", torch.tensor([2, 1, 5, 1, 3])),
-        (2, _mc_k_preds, _mc_k_target, "micro", torch.tensor([3, 3, 3, 0, 3])),
+        # (2, _mc_k_preds, _mc_k_target, "micro", torch.tensor([3, 3, 3, 0, 3])),
```
Comment (lantiga) [inline tests/unittests/classification/test_stat_scores.py:417]: remove comment

File: tests/unittests/classification/test_stat_scores.py [hunk 9]
```diff
@@ -361,27 +362,85 @@ def test_raises_error_on_too_many_classes(preds, target, ignore_index, error_mes
...
+        (2, _mc_k_preds, _mc_k_target, "micro", torch.tensor([3, 0, 6, 0, 3])),
         (1, _mc_k_preds, _mc_k_target, None, torch.tensor([[0, 1, 1], [0, 1, 0], [2, 1, 2], [1, 0, 0], [1, 1, 1]])),
-        (2, _mc_k_preds, _mc_k_target, None, torch.tensor([[1, 1, 1], [1, 1, 1], [1, 1, 1], [0, 0, 0], [1, 1, 1]])),
+        # (2, _mc_k_preds, _mc_k_target, None, torch.tensor([[1, 1, 1], [1, 1, 1], [1, 1, 1], [0, 0, 0], [1, 1, 1]])),
```
Comment (lantiga) [inline tests/unittests/classification/test_stat_scores.py:420]: remove comment

File: tests/unittests/classification/test_specificity.py [hunk 10]
```diff
@@ -354,24 +354,50 @@ def test_multiclass_specificity_dtype_gpu(self, inputs, dtype):
...
+])
 
 
 @pytest.mark.parametrize(
     ("k", "preds", "target", "average", "expected_spec"),
     [
-        (1, _mc_k_preds, _mc_k_target, "micro", tensor(5 / 6)),
-        (2, _mc_k_preds, _mc_k_target, "micro", tensor(1 / 2)),
+        (1, _mc_k_preds, _mc_k_target, "micro", torch.tensor(5 / 6)),
+        # (2, _mc_k_preds, _mc_k_target, "micro", tensor(1 / 2)),
```
Comment (lantiga) [inline tests/unittests/classification/test_specificity.py:373]: remove comment

**Review by lantiga — APPROVED** [review]

**Comment by lantiga** [comment]
Thanks @rittik9!

File: tests/unittests/classification/test_specificity.py (hunk 10 above)
Comment (Borda) [reply tests/unittests/classification/test_specificity.py:373]:
```suggestion
        (2, _mc_k_preds, _mc_k_target, "micro", tensor(1 / 2)),
```

File: tests/unittests/classification/test_precision_recall.py [hunk 11]
```diff
@@ -394,10 +402,25 @@ def test_multiclass_precision_recall_half_gpu(self, inputs, module, functional,
 @pytest.mark.parametrize(
     ("k", "preds", "target", "average", "expected_prec", "expected_recall"),
     [
-        (1, _mc_k_preds, _mc_k_target, "micro", tensor(2 / 3), tensor(2 / 3)),
-        (2, _mc_k_preds, _mc_k_target, "micro", tensor(1 / 2), tensor(1.0)),
-        (1, _mc_k_preds2, _mc_k_targets2, "macro", tensor(1 / 3), tensor(1 / 2)),
-        (2, _mc_k_preds2, _mc_k_targets2, "macro", tensor(1 / 3), tensor(1 / 2)),
+        (1, _mc_k_preds, _mc_k_target, "micro", torch.tensor(2 / 3), torch.tensor(2 / 3)),
+        # (2, _mc_k_preds, _mc_k_target, "micro", tensor(1 / 2), tensor(1.0)),
```
Comment (Borda) [inline tests/unittests/classification/test_precision_recall.py:406]:
```suggestion
        (2, _mc_k_preds, _mc_k_target, "micro", tensor(1 / 2), tensor(1.0)),
```

File: tests/unittests/classification/test_stat_scores.py (hunk 8 above)
Comment (Borda) [reply tests/unittests/classification/test_stat_scores.py:417]:
```suggestion
        (2, _mc_k_preds, _mc_k_target, "micro", torch.tensor([3, 3, 3, 0, 3])),
```

File: tests/unittests/classification/test_stat_scores.py (hunk 9 above)
Comment (Borda) [reply tests/unittests/classification/test_stat_scores.py:420]:
```suggestion
        (2, _mc_k_preds, _mc_k_target, None, torch.tensor([[1, 1, 1], [1, 1, 1], [1, 1, 1], [0, 0, 0], [1, 1, 1]])),
```

File: tests/unittests/classification/test_stat_scores.py (hunk 8 above)
Comment (Borda) [reply tests/unittests/classification/test_stat_scores.py]: @rittik9 so they were commented out because they were wrong?

Comment (rittik9) [reply tests/unittests/classification/test_stat_scores.py]: Yes, I added new corrected  values for the input configurations.

Comment (Borda) [reply tests/unittests/classification/test_stat_scores.py:418]: Ok, then delete the commented lines

File: tests/unittests/classification/test_f_beta.py [hunk 12]
```diff
@@ -391,7 +403,34 @@ def test_multiclass_fbeta_score_half_gpu(self, inputs, module, functional, compa
     ("k", "preds", "target", "average", "expected_fbeta", "expected_f1"),
     [
         (1, _mc_k_preds, _mc_k_target, "micro", torch.tensor(2 / 3), torch.tensor(2 / 3)),
-        (2, _mc_k_preds, _mc_k_target, "micro", torch.tensor(5 / 6), torch.tensor(2 / 3)),
+        # (2, _mc_k_preds, _mc_k_target, "micro", torch.tensor(5 / 6), torch.tensor(2 / 3)),
```
Comment (Borda) [inline tests/unittests/classification/test_f_beta.py:406]:
```suggestion
```

File: tests/unittests/classification/test_precision_recall.py [hunk 13]
```diff
@@ -395,10 +403,25 @@ def test_multiclass_precision_recall_half_gpu(self, inputs, module, functional,
...
```
Comment (Borda) [inline tests/unittests/classification/test_precision_recall.py:407]:
```suggestion
```

File: tests/unittests/classification/test_specificity.py [hunk 14]
```diff
@@ -355,24 +355,50 @@ def test_multiclass_specificity_dtype_gpu(self, inputs, dtype):
...
```
Comment (Borda) [inline tests/unittests/classification/test_specificity.py:374]:
```suggestion
```

File: tests/unittests/classification/test_stat_scores.py [hunk 15]
```diff
@@ -362,27 +363,85 @@ def test_raises_error_on_too_many_classes(preds, target, ignore_index, error_mes
...
+    [0.3, 0.3, 0.4],
+])
+
 
 @pytest.mark.parametrize(
     ("k", "preds", "target", "average", "expected"),
     [
         (1, _mc_k_preds, _mc_k_target, "micro", torch.tensor([2, 1, 5, 1, 3])),
-        (2, _mc_k_preds, _mc_k_target, "micro", torch.tensor([3, 3, 3, 0, 3])),
+        # (2, _mc_k_preds, _mc_k_target, "micro", torch.tensor([3, 3, 3, 0, 3])),
```
Comment (Borda) [inline tests/unittests/classification/test_stat_scores.py:418]:
```suggestion
```

File: tests/unittests/classification/test_stat_scores.py [hunk 16]
```diff
@@ -362,27 +363,85 @@ def test_raises_error_on_too_many_classes(preds, target, ignore_index, error_mes
...
+        (2, _mc_k_preds, _mc_k_target, "micro", torch.tensor([3, 0, 6, 0, 3])),
         (1, _mc_k_preds, _mc_k_target, None, torch.tensor([[0, 1, 1], [0, 1, 0], [2, 1, 2], [1, 0, 0], [1, 1, 1]])),
-        (2, _mc_k_preds, _mc_k_target, None, torch.tensor([[1, 1, 1], [1, 1, 1], [1, 1, 1], [0, 0, 0], [1, 1, 1]])),
+        # (2, _mc_k_preds, _mc_k_target, None, torch.tensor([[1, 1, 1], [1, 1, 1], [1, 1, 1], [0, 0, 0], [1, 1, 1]])),
```
Comment (Borda) [inline tests/unittests/classification/test_stat_scores.py:421]:
```suggestion
```
